### PR TITLE
Morpho Aave v3 resolver

### LINKED ISSUE #104: Resume Morpho Resolver
- [x] Morpho Aave V3 contracts
 - [x] Testcases

Reference material -> 
SDK - [https://www.npmjs.com/package/@morpho-org/morpho-aave-v3-sdk](https://www.npmjs.com/package/@morpho-org/morpho-aave-v3-sdk)

Docs - [https://developers.morpho.xyz/track-and-manage-position/track-positions-on-morpho/track-supply-positions](https://developers.morpho.xyz/track-and-manage-position/track-positions-on-morpho/track-supply-positions)

Snippets Repo - [https://github.com/morpho-org/morpho-snippets](https://github.com/morpho-org/morpho-snippets)

### PR DESCRIPTION
**Pull request by q1q0**
Closes #104

### CODE REVIEW

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    function getAWethAddr() internal pure returns (address) {
+        return 0x030bA81f1c18d280636F32af80b9AAd02Cf0854e;
+    }
+
+    function getChainlinkEthFeed() internal pure returns (address) {
+        return 0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419; // should be updated after deployment
+    }
+
+    function getAaveProtocolDataProvider() internal pure returns (address) {
+        return 0x057835Ad21a177dbdd3090bB1CAE03EaCF78Fc6d; // should be updated after deployment
```
Comment (shriyatyagii) [inline contracts/protocols/mainnet/morpho/aave_v3/helpers.sol]: We need Aave v3 addresses, not Aave v2.
Update from here ->
https://docs.aave.com/developers/deployed-contracts/v3-mainnet/ethereum-mainnet

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    function getAaveIncentivesController() internal pure returns (address) {
+        return 0xd784927Ff2f95ba542BfC824c8a8a98F3495f6b5; // should be updated after deployment
```
Comment (shriyatyagii) [inline contracts/protocols/mainnet/morpho/aave_v3/helpers.sol]: Update this as well

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    }
+
+    struct MorphoData {
+        MarketDetail[] aaveMarketsCreated;
+        bool isClaimRewardsPausedAave;
+        uint128 repay;
+        uint128 withdraw;
```
Comment (shriyatyagii) [inline contracts/protocols/mainnet/morpho/aave_v3/helpers.sol]: We can remove these two

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+
+    function getUserMarketData(
+        address user,
+        address underlying,
+        uint256 priceInEth,
+        uint256 priceInUsd
+    ) internal view returns (UserMarketData memory userMarketData_) {
+        userMarketData_.marketData = getMarketData(underlying, priceInEth, priceInUsd);
+
+        userMarketData_.p2pBorrows = morphoGetter.scaledP2PBorrowBalance(underlying, user);
```
Comment (shriyatyagii) [inline contracts/protocols/mainnet/morpho/aave_v3/helpers.sol]: We need to convert the scaled values to get the underlying values.
If you see, they are using `_getUserSupplyBalanceFromIndexes` inside `supplyBalance` to get the main balance. You can take the calculation logic from here.

image

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+        userMarketData_.poolBorrows = morphoGetter.scaledPoolBorrowBalance(underlying, user);
```
Comment (shriyatyagii) [inline contracts/protocols/mainnet/morpho/aave_v3/helpers.sol]: Same for this, and all the other scaled values.

File: contracts/protocols/mainnet/morpho/aave_v3/helpers.sol (hunk 4 above)
Comment (q1q0) [reply contracts/protocols/mainnet/morpho/aave_v3/helpers.sol:254]: To get a logic of rayMulDown we should access here.
https://github.com/morpho-labs/morpho-data-structures-private/tree/06a9fb3fdf0d089977d99e5cd62774fb79f603a7
But I can't do it now.

**Pull request closed without merging**